pkg/upstream/cluster/loadbalancer.go

```diff
@@ -152,80 +152,49 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
-	hosts         types.HostSet
-	hostsWeighted []*hostSmoothWeighted
-	lock          sync.Mutex
-}
-
-type hostSmoothWeighted struct {
-	weight          int64
-	currentWeight   int64
-	effectiveWeight int64
-}
-
-func newSmoothWeightedRRLoadBalancer(info types.ClusterInfo, hosts types.HostSet) types.LoadBalancer {
-	smoothWRRLoadBalancer := &smoothWeightedRRLoadBalancer{
-		hosts:         hosts,
-		hostsWeighted: make([]*hostSmoothWeighted, len(hosts.Hosts())),
-	}
-	// iterate over all hosts to init host with Weighted
-	for idx, host := range hosts.Hosts() {
-		w := int64(host.Weight())
-		smoothWRRLoadBalancer.hostsWeighted[idx] = &hostSmoothWeighted{
-			effectiveWeight: w,
-			weight:          w,
-		}
-	}
-	return smoothWRRLoadBalancer
+type WeightedRRLoadBalancer struct {
+	*EdfLoadBalancer
+	rrIndex uint32
 }
 
-// smooth weighted round robin
-// O(n), traverse over all hosts
-func (lb *smoothWeightedRRLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
-	var totalWeight int64 = 0
-	var selectedHostWeighted *hostSmoothWeighted
-	var selectedHost types.Host
-	for idx, host := range lb.hosts.Hosts() {
-		if !host.Health() {
-			continue
-		}
-		hw := lb.hostsWeighted[idx]
-		atomic.AddInt64(&hw.currentWeight, atomic.LoadInt64(&hw.effectiveWeight))
-		totalWeight += atomic.LoadInt64(&hw.effectiveWeight)
-
-		if hw.effectiveWeight < hw.weight {
-			atomic.AddInt64(&hw.effectiveWeight, 1)
-		}
-
-		if selectedHostWeighted == nil || atomic.LoadInt64(&hw.currentWeight) > atomic.LoadInt64(&selectedHostWeighted.currentWeight) {
-			selectedHostWeighted = hw
-			selectedHost = host
-		}
+func newWeightedRRLoadBalancer(info types.ClusterInfo, hosts types.HostSet) types.LoadBalancer {
+	rrLB := &WeightedRRLoadBalancer{
 	}
-	//
-	if selectedHostWeighted != nil {
-		atomic.AddInt64(&selectedHostWeighted.currentWeight, -totalWeight)
+	rrLB.EdfLoadBalancer = newEdfLoadBalancerLoadBalancer(hosts, rrLB.unweightChooseHost, rrLB.hostWeight)
+	var idx uint32
+	hostsList := hosts.Hosts()
+	rrLB.mutex.Lock()
+	defer rrLB.mutex.Unlock()
+	if len(hostsList) != 0 {
+		idx = rrLB.rand.Uint32() % uint32(len(hostsList))
 	}
-	return selectedHost
+	rrLB.rrIndex = idx
+	return rrLB
 }
 
-func (lb *smoothWeightedRRLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCriteria) bool {
+func (lb *WeightedRRLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCriteria) bool {
 	return len(lb.hosts.Hosts()) > 0
 }
 
-func (lb *smoothWeightedRRLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) int {
+func (lb *WeightedRRLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) int {
 	return len(lb.hosts.Hosts())
 }
 
+func (lb *WeightedRRLoadBalancer) hostWeight(item WeightItem) float64 {
+	host := item.(types.Host)
+	return float64(host.Weight())
+}
+// do unweighted (fast) selection
+func (lb *WeightedRRLoadBalancer) unweightChooseHost(context types.LoadBalancerContext) types.Host {
+	targets := lb.hosts.Hosts()
+	total := len(targets)
```

Comment:
need check value of `total`, if it's 0 should return nil.

Reply:
Already judged in the `EdfLoadbalancer#ChooseHost`.